File: check.py
import asyncio
import os
import time
from datetime import datetime

from db import Settings, engines, sessions

PYPPETEER_CHROMIUM_REVISION = '1263111'
os.environ['PYPPETEER_CHROMIUM_REVISION'] = PYPPETEER_CHROMIUM_REVISION

from pyppeteer import launch

import nest_asyncio

nest_asyncio.apply()

script_dir = os.path.dirname(os.path.abspath(__file__))

async def html_to_pdf(html_content):
    browser = await launch(handleSIGINT=False,
                           handleSIGTERM=False,
                           handleSIGHUP=False)
    pages = await browser.pages()

    page = pages[0]
    await page.setContent(html_content)

    # Set the viewport size to ensure content is rendered
    await page.setViewport({'width': 1920, 'height': 1080})

    # Wait for any asynchronous content to load
    await page.waitForSelector("#p1")
    time.sleep(2)
    # Get the dimensions of the entire page
    dimensions = await page.evaluate('''() => {
        const element = document.getElementById('p1');
        return {
            width: element.offsetWidth,
            height: element.offsetHeight
        };
    }''')

    await page.pdf({
        'title': 'чек',
        'path': os.path.join(script_dir, 'check.pdf'),
        'width': dimensions['width'],
        'height': dimensions['height'],
        'printBackground': True,
        'margin': {
            'top': '0px',
            'right': '0px',
            'bottom': '0px',
            'left': '0px'
        }
    })

    await browser.close()


def get_check(bank, d):
    data = {
               'DOCUMENT': '4252228616',
               'SENDER_CARD': '2314',
               'RECEIVER_CARD': '2511',
               'CODE': '212251',
               'AMOUNT': '1234.13',
               'TIME': datetime.now().strftime('%d.%m.%Y %H:%M'),
               'RECEIVER_NAME': 'Скам Скамыч Ф.',
               'RECEIVER_NUMBER': '+7 (993) 111-22-33',
           } | d
    with open(os.path.join(script_dir, bank, '1.html'), 'r', encoding='utf-8') as file:
        html_content = file.read()
    for key, value in data.items():
        html_content = html_content.replace(key, str(value))
    html_content = html_content.replace('$URL', 'http://127.0.0.1:5000/' + bank)
    loop = asyncio.get_event_loop()
    loop.run_until_complete(html_to_pdf(html_content))
    
    from PyPDF2 import PdfReader, PdfWriter
    from datetime import datetime

    pdf_path = os.path.join(script_dir, 'check.pdf')
    reader = PdfReader(pdf_path)
    writer = PdfWriter()

    for page in reader.pages:
        writer.add_page(page)

    writer.add_metadata({
        '/Title': 'Квитанция-11100399',
        '/Author': '',
        '/Subject': '"/reports/IB/Receipt"',
        '/Keywords': '22.10.2024 08:41:12|f8848b21-135a-4ad6-b014-15c00b30e714|7258',
        '/Producer': 'OpenPDF 1.3.30.jaspersoft.2',
        '/Creator': 'JasperReports Library version 6.20.3-415f9428cffdb6805c6f85bbb29ebaf18813a2ab',
        '/CreationDate': datetime(2024, 10, 22, 8, 41, 12),
        '/ModDate': datetime(2024, 10, 22, 8, 41, 12)
    })

    with open(pdf_path, 'wb') as output_file:
        writer.write(output_file)

    return open(pdf_path, 'rb').read()


# get_check('tinkoff', {'SENDER_NAME': 'Егор Иванович Ж.', 'RECEIVER_NAME': 'Алексей Михайлович З.', "AMOUNT": '59664.10'})

File: netlify/functions/check.js

import puppeteer from 'puppeteer-core';
import fs from 'fs';
import path from 'path';

const scriptDir = ''
import {executablePath} from 'puppeteer'

export default async (req) => {
  const { bank, data } = await req.json();

  try {
    const htmlContent = await getHtmlContent(bank, data);
    const pdfBuffer = await htmlToPdf(htmlContent);

    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/pdf',
        'Content-Disposition': 'attachment; filename=check.pdf'
      },
      body: pdfBuffer.toString('base64'),
      isBase64Encoded: true
    };
  } catch (error) {
    console.error('Error:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal Server Error' })
    };
  }
};

async function getHtmlContent(bank, data) {
  let htmlContent = await fs.promises.readFile(path.join(scriptDir, bank, '1.html'), 'utf-8');
  
  for (const [key, value] of Object.entries(data)) {
    htmlContent = htmlContent.replaceAll(key, String(value));
  }
  htmlContent = htmlContent.replaceAll('$URL', 'http://127.0.0.1:5000/' + bank);

  return htmlContent;
}

async function htmlToPdf(htmlContent) {
  const browser = await puppeteer.launch({
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
    ignoreHTTPSErrors: true,
    executablePath: executablePath(),
    headless: false
  });

  try {
    const page = (await browser.pages())[0];
    await page.setContent(htmlContent);
    await page.setViewport({ width: 1920, height: 1080 });

    await page.waitForSelector("#p1");
    await new Promise(resolve => setTimeout(resolve, 2000));

    const dimensions = await page.evaluate(() => {
      const element = document.getElementById('p1');
      return element ? {
        width: element.offsetWidth,
        height: element.offsetHeight
      } : null;
    });

    if (!dimensions) {
      throw new Error('Failed to get dimensions');
    }

    const pdfBuffer = await page.pdf({
      width: dimensions.width,
      height: dimensions.height,
      printBackground: true,
      margin: {
        top: '0px',
        right: '0px',
        bottom: '0px',
        left: '0px'
      }
    });

    return pdfBuffer;
  } finally {
    await browser.close();
  }
}
const currentTime = new Date().toLocaleString('ru-RU', { day: '2-digit', month: '2-digit', year: '2-digit', hour: '2-digit', minute: '2-digit' });
const htmlContent = await getHtmlContent('tinkoff', {'TIME': currentTime, 'SENDER_NAME': 'Егор Иванович Ж.', 'RECEIVER_NUMBER':'+7 (912) 999-55-66','RECEIVER_NAME': 'Алексей Михайлович З.', "AMOUNT": '59664.10'});
const pdfBuffer = await htmlToPdf(htmlContent);
import { PDFDocument } from 'pdf-lib'
const pdf = await PDFDocument.load(pdfBuffer);
pdf.setTitle('Квитанция-11100399')
pdf.setAuthor('')
pdf.setSubject('"/reports/IB/Receipt"')
pdf.setKeywords(['22.10.2024', '08:41:12|f8848b21-135a-4ad6-b014-15c00b30e714|7258'])
pdf.setProducer('OpenPDF 1.3.30.jaspersoft.2')
pdf.setCreator('JasperReports Library version 6.20.3-415f9428cffdb6805c6f85bbb29ebaf18813a2ab')
pdf.setCreationDate(new Date('2024-10-22 08:41:12+03:00'))
pdf.setModificationDate(new Date('2024-10-22 08:41:12+03:00'))

fs.writeFileSync('tinkoff.pdf', await pdf.save({useObjectStreams: false}));

